xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc

```diff
@@ -0,0 +1,1708 @@
+ public:
+  TranslationState() = default;
+
+  absl::Status setFileName(::xls::Fileno num, StringAttr name) {
+    if (fileno_map_.contains(num)) {
+      return absl::InternalError(
+          absl::StrCat("Duplicate file number ", num.value()));
+    }
+    fileno_map_[num] = name;
+    return absl::OkStatus();
+  }
+
+  absl::StatusOr<StringAttr> getFileName(::xls::Fileno no) {
+    auto op = fileno_map_.find(no);
+    if (op == fileno_map_.end()) {
+      return absl::InternalError(
+          absl::StrCat("Unknown file number  ", no.value()));
+    } else {
+      return op->second;
+    }
+  }
+
+  absl::Status setFn(std::string name, FlatSymbolRefAttr fn) {
+    if (fn_map_.contains(name)) {
+      return absl::InternalError(
+          absl::StrCat("Duplicate function name ", name));
+    }
+    fn_map_[name] = fn;
+    return absl::OkStatus();
+  }
+
+  absl::StatusOr<FlatSymbolRefAttr> getFn(std::string name) {
+    auto op = fn_map_.find(name);
+    if (op == fn_map_.end()) {
+      return absl::InternalError(absl::StrCat("Unknown function name  ", name));
+    } else {
+      return op->second;
+    }
+  }
+
+  absl::Status setChn(std::string name, FlatSymbolRefAttr chn) {
+    if (chn_map_.contains(name)) {
+      return absl::InternalError(absl::StrCat("Duplicate channel name ", name));
+    }
+    chn_map_[name] = chn;
+    return absl::OkStatus();
+  }
+
+  absl::StatusOr<SymbolRefAttr> getChn(std::string name) {
+    auto op = chn_map_.find(name);
+    if (op == chn_map_.end()) {
+      return absl::InternalError(absl::StrCat("Unknown channel name  ", name));
+    } else {
+      return op->second;
+    }
+  }
+
+  absl::Status setMlirValue(int64_t id, Value op) {
+    if (ssa_map_.contains(id)) {
+      return absl::InternalError(
+          absl::StrCat("Duplicate assignment to value id ", id));
+    }
+    ssa_map_[id] = op;
+    return absl::OkStatus();
+  }
+
+  absl::StatusOr<Value> getMlirValue(int64_t id) {
+    auto op = ssa_map_.find(id);
+    if (op == ssa_map_.end()) {
+      return absl::InternalError(absl::StrCat("Unknown value id ", id));
+    } else {
+      return op->second;
+    }
+  }
+
+  void newFunctionBaseContext(bool is_proc) {
+    this->ssa_map_ = {};
+    this->is_proc_ = is_proc;
+    this->state_elem_map = {};
+  }
+
+  void setStateElem(std::string name, Value elem) {
+    assert(is_proc_ && "Only proc functions contain state elements");
+    state_elem_map[name] = elem;
+  }
+
+  absl::StatusOr<Value> getStateElem(std::string id) {
+    assert(is_proc_ && "Only proc functions contain state elements");
+    auto elem = state_elem_map.find(id);
+    if (elem == state_elem_map.end()) {
+      return absl::InternalError(absl::StrCat("Unknown state elem ", id));
+    } else {
+      return elem->second;
+    }
+  }
+
+ private:
+  // --- package/module state ----
+  // Maps name -> MLIR ref for all functions in current
+  // package/module.
+  std::unordered_map<std::string, FlatSymbolRefAttr> fn_map_;
```

Comment:
given you're using absl already, can you use absl::flat_hash_map instead of std::unordered_map everywhere please? absl::flat_hash_map has the same API as unordered_map but is significantly more performant.